### PR TITLE
Sync inventory with server side after sell items to shop

### PR DESCRIPTION
**Pull request by Zaklinatel**
Fixed: If the mod is installed only on the server side, sold items not removed from the player’s inventory in case items not enough in a hand.

### CODE REVIEW

File: src/main/java/com/forgeessentials/economy/shop/ShopManager.java [hunk 1]
```diff
@@ -305,16 +305,21 @@ public void playerInteractEvent(final PlayerInteractEvent event)
...
                 removedAmount = Math.min(equippedStack.stackSize, transactionStack.stackSize);
                 equippedStack.stackSize -= removedAmount;
                 if (equippedStack.stackSize <= 0)
                     event.entityPlayer.inventory.mainInventory[event.entityPlayer.inventory.currentItem] = null;
             }
-            if (removedAmount < transactionStack.stackSize)
-                removedAmount += ModuleEconomy.tryRemoveItems(event.entityPlayer, transactionStack, transactionStack.stackSize - removedAmount);
+
+            if (removedAmount < transactionStack.stackSize) {
+                ModuleEconomy.tryRemoveItems(event.entityPlayer, transactionStack, transactionStack.stackSize - removedAmount);
```
Comment (spacebuilder2020) [inline src/main/java/com/forgeessentials/economy/shop/ShopManager.java:320]: What is the reason for this change?

File: src/main/java/com/forgeessentials/economy/ModuleEconomy.java [hunk 2]
```diff
@@ -177,25 +177,28 @@ public static void confirmNewWalletAmount(UserIdent ident, Wallet wallet)
...
+                    player.inventory.setInventorySlotContents(slot, null);
+                } else {
+                    player.inventory.decrStackSize(slot, removeCount);
+                }
+
+                amount -= removeCount;
             }
         }
+
+        player.inventoryContainer.detectAndSendChanges();
```
Comment (spacebuilder2020) [inline src/main/java/com/forgeessentials/economy/ModuleEconomy.java:200]: Isn't this the only thing needed to sync with client side.  What is the reason for changing the rest of the logic?

File: src/main/java/com/forgeessentials/economy/shop/ShopManager.java (hunk 1 above)
Comment (Zaklinatel) [reply src/main/java/com/forgeessentials/economy/shop/ShopManager.java:320]: removedAmount variable is not used in the next code, so there is no point in the last change. Needless operation.

File: src/main/java/com/forgeessentials/economy/ModuleEconomy.java (hunk 2 above)
Comment (Zaklinatel) [reply src/main/java/com/forgeessentials/economy/ModuleEconomy.java:200]: Please note that there were two identical loops that contained almost identical code with the only difference one loop counted the number of items, and the other deleted them. But it can be easily done in one loop.
Also I added `amount > 0` to loop exit condition on case the required number of items is removed.
Just small optimization.
(sorry for my english)

Comment (spacebuilder2020) [reply src/main/java/com/forgeessentials/economy/ModuleEconomy.java:200]: The problem with optimizations is it requires additional testing as I don't know your code well enough to know if you broke existing functionality when adding your optimizations.  Also, please note that in the pull request you only stated that you fixed a client desync bug.  I have a policy of not accepting pulls that go beyond the stated scope.  Also, for a simple fix, (stated scope), how come you did not make a bug report?

File: src/main/java/com/forgeessentials/economy/shop/ShopManager.java (hunk 1 above)
Comment (spacebuilder2020) [reply src/main/java/com/forgeessentials/economy/shop/ShopManager.java:320]: This is an unnecessary change.  While it is true that removedAmount is not being used, there was no need to make this change as you did not change any fundamental logic and this just adds commit bloat.

File: src/main/java/com/forgeessentials/economy/ModuleEconomy.java (hunk 2 above)
Comment (Zaklinatel) [reply src/main/java/com/forgeessentials/economy/ModuleEconomy.java:200]: > I have a policy of not accepting pulls that go beyond the stated scope

Well, I understand your position. You're right. I really allowed myself to make more changes than required by the bugfix. You can just throw out all unnecessary changes and keep only `player.inventoryContainer.detectAndSendChanges();` - that's will be enough.

> Also, for a simple fix, (stated scope), how come you did not make a bug report?

This mod is currenty running on my server and I had to urgently fix this bug without waiting for someone else to do it.

File: src/main/java/com/forgeessentials/economy/shop/ShopManager.java (hunk 1 above)
Comment (Zaklinatel) [reply src/main/java/com/forgeessentials/economy/shop/ShopManager.java:320]: Ok, you're right. My bad.

File: src/main/java/com/forgeessentials/economy/ModuleEconomy.java (hunk 2 above)
Comment (spacebuilder2020) [reply src/main/java/com/forgeessentials/economy/ModuleEconomy.java:200]: That is understandable.  However, making an issue is still good as it allows us to keep track of the bug.  I was unaware that this bug even existed.

Comment (spacebuilder2020) [reply src/main/java/com/forgeessentials/economy/ModuleEconomy.java:200]: Btw, in the future, if you have an issue, first contact us on discord, then if we can't solve it their make a bug report.  If we don't have time to make the fix and you indicate you have an idea of a fix, then we will suggest you make a PR.   It just helps this stay organized.  I will look at and test your optimizations after work today.

Comment (Zaklinatel) [reply src/main/java/com/forgeessentials/economy/ModuleEconomy.java:200]: Yes, I understand. For some reason I didn't think at all about any organization. Of course it was a mistake.

Moreover, I just noticed I left a new error in the code: I forgot about the return value of `amount` variable. It no longer returns how many items have been removed from inventory. Please check it when you will reviewing my commit.

Comment (spacebuilder2020) [reply src/main/java/com/forgeessentials/economy/ModuleEconomy.java:200]: Go ahead and make these fixes as an additional commit.  Make sure to test everything and verify that it is working.  Your logic looks sound for the other stuff though.

**Review by spacebuilder2020 — APPROVED** [review]
I am going to go ahead and merge this so it does not remain hanging out too long.

I will then fix the issue with amount in another commit.

**Comment by Zaklinatel** [comment]
Ok.
